src/main/java/net/minestom/server/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,362 @@
+        });
+        return result;
+    }
+
+    /**
+     * Checks whether {@link ItemStack} can be fully taken from the inventory.
+     *
+     * @param itemStack the item to be checked
+     * @return true if the item can be fully taken from the inventory, false otherwise
+     */
+    public boolean canTakeItemStack(@NotNull ItemStack itemStack) {
+        final StackingRule stackingRule = itemStack.getStackingRule();
+        int amountLeft = stackingRule.getAmount(itemStack);
+        for (int i = 0; i < getInnerSize(); i++) {
+            ItemStack inventoryItem = getItemStack(i);
+            if (inventoryItem.isAir()) {
+                continue;
+            }
+            if (stackingRule.canBeStacked(itemStack, inventoryItem)) {
+                final int itemAmount = stackingRule.getAmount(inventoryItem);
+                if (amountLeft <= itemAmount) {
+                    return true;
+                }
+                amountLeft -= itemAmount;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether {@link ItemStack}s can be fully taken from the inventory.
+     *
+     * @param itemStacks items to be checked
+     * @return true if all the items can be fully taken from the inventory, false otherwise
+     */
+    public boolean canTakeItemStacks(@NotNull List<ItemStack> itemStacks) {
+        return itemStacks.stream().allMatch(this::canTakeItemStack);
+    }
+
+    public synchronized void replaceItemStack(int slot, @NotNull UnaryOperator<@NotNull ItemStack> operator) {
+        var currentItem = getItemStack(slot);
+        setItemStack(slot, operator.apply(currentItem));
+    }
+
+    /**
+     * Clears the inventory and send relevant update to the viewer(s).
+     */
+    public synchronized void clear() {
+        // Clear the item array
+        Arrays.fill(itemStacks, ItemStack.AIR);
+        // Send the cleared inventory to viewers
+        update();
+    }
+
+    public abstract void update();
+
+    /**
+     * Gets the {@link ItemStack} at the specified slot.
+     *
+     * @param slot the slot to check
+     * @return the item in the slot {@code slot}
+     */
+    public @NotNull ItemStack getItemStack(int slot) {
+        return itemStacks[slot];
+    }
+
+    /**
+     * Gets all the {@link ItemStack} in the inventory.
+     * <p>
+     * Be aware that the returned array does not need to be the original one,
+     * meaning that modifying it directly may not work.
+     *
+     * @return an array containing all the inventory's items
+     */
+    public @NotNull ItemStack[] getItemStacks() {
+        return itemStacks.clone();
+    }
+
+    /**
+     * Gets the size of the inventory.
+     *
+     * @return the inventory's size
+     */
+    public int getSize() {
+        return size;
+    }
+
+    /**
+     * Gets the size of the "inner inventory" (which includes only "usable" slots).
+     *
+     * @return inner inventory's size
+     */
+    public int getInnerSize() {
```

Comment:
Could this be renamed to `getUsableSlots` or something that represents better what this method actually returns?

Reply:
I feel that this should be replaced by a kind of Inventory Section API